designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,556 @@
+}
+```
+
+The context will be responsible for managing the memory of any data  
+structures in LatentData, such as `Eigen::Matrix` or `std::vector`
+instances. By reusing the same `LatentData` variable (per thread),
+variables can be reset rather than allocating and freeing each evaluation.
+
+
+### Modified log_prob method
+
+The existing `log_prob` method must be updated to maintain backward
+compatibility: 
+
+```cpp
+template <bool propto__, bool jacobian__, typename T_> inline T_
+log_prob(Eigen::Matrix<T_,-1,1>& params_r, 
+         const LatentData& latent_data__,
+         std::ostream* pstream = nullptr) const;
+```
+
+In this case, the `log_prob` code generation must be updated to copy
+all of the generated data into scope.
+
+```cpp
+template <bool propto__, bool jacobian__, typename T_> inline T_ 
+log_prob(
+    Eigen::Matrix<T_,-1,1>& params_r,  
+    const LatentData& latent_data  , 
+    std::ostream* pstream = nullptr) const {
+  double sens = latent_data__.sens;
+  double spec = latent_data__.spec;
+  ... code generate as before with sens/spec in scope ...
+}
+```
+
+For ease of backward compatibility, we can have the old implementation
+with a default value,
+
+```cpp  
+template <bool propto__, bool jacobian__, typename T_> inline T_ 
+log_prob(
+    Eigen::Matrix<T_,-1,1>& params_r,  
+    std::ostream* pstream = nullptr) const {
+  LatentData latent_data,
+  return log_prob(params_r, latent_data, pstream) const;
+```
+
+### BridgeStan
+
+BridgeStan can be updated to mimic the updates to the Stan model
+class.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. It will require additional documentation and tests, increasing
+   Stan's long-term maintenance burden and the overhead required for
+   someone to understand what it's doing. 
+2. Using "cut"-based inference does not perform full Bayesian
+   inference, and there's a risk that this will confuse our users. 
+3. Using this block for Gibbs sampling will be less efficient than
+   marginalization and may tempt many users to define poorly mixing
+   models. 
+4. There is nothing enforcing consistency of the conditional
+   distributions when doing Gibbs, so this will be easy to get wrong
+   and there won't be an easy way to test that this is right.
+5.  Introducing a `latent data` block opens up a back door approach to
+    converting real numbers to integers, which can be done with
+    `to_int(data real)` if the argument is "data" (i.e., a primitive
+    rather than autodiff variable). 
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## Alternatives
+
+To support pure uncertainty propagation and non-interacting forms of
+cut (i.e., ones that don't use any of the model parameters), we can
+use multiple imputation in the traditional way by running multiple
+Stan programs in sequence. This is less efficient and not as general.
+
+There is no alternative for the Gibbs-based discrete parameter
+inference. 
```

Comment:
The first alternative is marginalization.
The second alternative is to implement the HMC+Gibbs algorithm in the Stan algorithms library and just allow models to declare discrete parameters.
Also we should consider special syntax for discrete parameters that may allow either marginalization or Gibbs chosen at runtime.

Reply:
I'll clarify that I meant there isn't an existing alternative for Gibbs within Stan.  

You can marginalize only when tractable, so it rules out high dimensional categorical and unbounded count data and also things like variable selection (not that those work well with Gibbs).

The original design of the model class included discrete parameters because I was thinking we could just use the entire density to update with Metropolis or with Metropolis-within-Gibbs style updates.

Reply:
Thanks---added both clarifications as suggested.  I'll leave this open for comment now rather than closing.

I think a proposal to add automatic Gibbs sampling to Stan should be another design document.  We could do generic Metropolis-within-Gibbs, but I asked the PyMC devs, and they say they can compute the Markov blanket to do this more efficiently.